event/parsing/totalBootTimeParser.go

```diff
@@ -0,0 +1,170 @@
+/**
+ *  Copyright (c) 2020  Comcast Cable Communications Management, LLC
+ */
+
+package parsing
+
+import (
+	"errors"
+	"regexp"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/goph/emperror"
+	"github.com/xmidt-org/glaukos/event/queue"
+	"github.com/xmidt-org/themis/xlog"
+)
+
+// TotalBootTimeParser takes fully-manageabke events and calculates the reboot time of a device by getting the last
+// reboot-pending event from codex.
+type TotalBootTimeParser struct {
+	Measures Measures
+	Logger   log.Logger
+	Client   EventClient
+	Label    string
+}
+
+var manageableRegex = regexp.MustCompile(".*/fully-manageable/")
+var rebootRegex = regexp.MustCompile(".*/reboot-pending/")
+
+/* Parse calculates boot time of devices by querying codex for the latest device events and performing
+ calculations. An analysis of codex events is only triggered by a fully-manageable event from caduceus.
+ Steps to calculate boot time:
+	 1) Determine if message is fully-manageable event.
+	 2) Get latest reboot-pending event from Codex where metadata field of /boot-time differs from the fully-manageable event.
+	 3) Subtract fully-manageable's birthdate from step 2's event birthdate.
+	 4) Record Metric.
+*/
+func (b TotalBootTimeParser) Parse(wrpWithTime queue.WrpWithTime) error {
+	// Add to metrics if no error calculating restart time.
+	if restartTime, err := b.calculateRestartTime(wrpWithTime); err == nil && restartTime > 0 {
+		b.Measures.TotalBootTimeHistogram.With(HardwareLabel, wrpWithTime.Message.Metadata[hardwareKey], FirmwareLabel, wrpWithTime.Message.Metadata[firmwareKey]).Observe(restartTime)
+	}
+
+	return nil
```

Comment:
This may not be in scope for this pr, but why do we always return nil? Is there a reason for the `Parse()` function returning an error?

Reply:
You're right, we should probably actually return an error if there's an error and log it or keep metrics on it. I've created an issue for it that I'll address in a later PR.